lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -320,6 +321,170 @@ void getAsmBlockArgumentNamesImpl(Operation *op, mlir::Region &region,
 static ParseResult parseNameKind(OpAsmParser &parser,
                                  firrtl::NameKindEnumAttr &result);
 
+//===----------------------------------------------------------------------===//
+// Layer Verification Utilities
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct CompareSymbolRefAttr {
+  // True if lhs is lexicographically less than rhs.
+  bool operator()(SymbolRefAttr lhs, SymbolRefAttr rhs) const {
+    auto x = lhs.getRootReference();
+    auto y = rhs.getRootReference();
+    switch (x.compare(y)) {
+    case -1:
+      return true;
+    case +1:
+      return false;
+    }
+    auto lhsNested = lhs.getNestedReferences();
+    auto rhsNested = rhs.getNestedReferences();
+    auto lhsNestedSize = lhsNested.size();
+    auto rhsNestedSize = rhsNested.size();
+    auto e = std::min(lhsNestedSize, rhsNestedSize);
+    for (unsigned i = 0; i < e; ++i) {
+      auto x = lhsNested[i].getAttr();
+      auto y = rhsNested[i].getAttr();
+      switch (x.compare(y)) {
+      case -1:
+        return true;
+      case +1:
+        return false;
+      }
+    }
+    return lhsNestedSize < rhsNestedSize;
+  }
+};
+} // namespace
+
+using LayerSet = SmallSet<SymbolRefAttr, 4, CompareSymbolRefAttr>;
+
+/// Get the ambient layers active at the given op.
+static LayerSet getAmbientLayersAt(Operation *op) {
+  // crawl through the parent ops, finding all ambient layers of
+  // the given operation.
+  LayerSet result;
+  for (; op != nullptr; op = op->getParentOp()) {
+    if (auto module = dyn_cast<FModuleLike>(op)) {
+      for (auto layer : module.getLayers())
+        result.insert(cast<SymbolRefAttr>(layer));
+      continue;
+    }
+    if (auto layerblock = dyn_cast<LayerBlockOp>(op)) {
+      result.insert(layerblock.getLayerName());
+      continue;
+    }
+  }
```

Comment:
> This is safe so long as all the ops are verified to be in a layerblock or a module (which is already checked!)

Just looked around in circt, I don't see this being verified for any expression op (including ref.sub, ref.send), nor most statement op (excepting layerblock, but including ref.define and when). We _are_ verifying the parent of declaration ops, though.

So, this code passes the verifiers today:
```mlir
builtin.module {
  %c = firrtl.constant 0  : !firrtl.uint<1>
  %p = firrtl.ref.send %c : !firrtl.uint<1>
}
```

I think it's reasonable to be a bit defensive here and leave the nullptr check in, but I will break when we hit the first module-like parent. Sound okay?

Reply:
Hmm... I'm surprised these checks aren't in there. There are some spotty checks that different FIRRTL ops have an `FModuleOp` parent, however, this is by no means consistent. I'd consider tightening this up so that all ops can only exist under `FModuleOp`, `WhenOp`, or `LayerBlockOp`. This may break other users or passes that are relying on this permissiveness.

Yes, this is fine as-is, then.

@nandor: are you aware of users needing to put FIRRTL operations in non-FIRRTL containers?